src/backend/bacon-video-widget-gst-missing-plugins.c

```diff
@@ -119,201 +72,105 @@ bacon_video_widget_gst_codec_install_plugin_is_blacklisted (const gchar * detail
-	gst_install_plugins_context_set_desktop_id (install_ctx, "org.gnome.Totem.desktop");
-	gst_install_plugins_context_set_confirm_search (install_ctx, confirm_search);
-	set_startup_notification_id (install_ctx);
-
-#ifdef GDK_WINDOWING_X11
-	display = gdk_display_get_default ();
-
-	if (GDK_IS_X11_DISPLAY (display) &&
-	    gtk_widget_get_window (GTK_WIDGET (ctx->bvw)) != NULL &&
-	    gtk_widget_get_realized (GTK_WIDGET (ctx->bvw)))
-	{
-		gulong xid = 0;
-
-		xid = bacon_video_widget_gst_get_toplevel (GTK_WIDGET (ctx->bvw));
-		gst_install_plugins_context_set_xid (install_ctx, xid);
+static CodecMissingSuggestedLaunchAction *
+determine_vlc_launch_action_for_missing_codecs (const gchar *mrl) {
+	GDesktopAppInfo *candidate;
+	const gchar *message_text;
+
+	/* First try org.videolan.VLC.desktop, and if it exists
+	 * use that, otherwise use eos-vlc.desktop */
+	candidate = g_desktop_app_info_new ("org.videolan.VLC.desktop");
+	if (candidate) {
+		message_text = _("This file type is currently unsupported in Videos. "
+		                 "video player. However, you can use VLC media player "
+		                 "to play it.");
+		return codec_missing_suggested_launch_action_new (message_text,
+								  _("Open with VLC"),
+								  candidate,
+								  mrl);
 	}
-#endif /* GDK_WINDOWING_X11 */
-
-	status = gst_install_plugins_async ((const char * const*) ctx->details, install_ctx,
-	                                    on_plugin_installation_done,
-	                                    ctx);
-
-	gst_install_plugins_context_free (install_ctx);
 
-	GST_INFO ("gst_install_plugins_async() result = %d", status);
-
-	if (status != GST_INSTALL_PLUGINS_STARTED_OK)
-	{
-		if (status == GST_INSTALL_PLUGINS_HELPER_MISSING)
-		{
-			g_message ("Automatic missing codec installation not supported "
-			           "(helper script missing)");
-		} else {
-			g_warning ("Failed to start codec installation: %s",
-			           gst_install_plugins_return_get_name (status));
-		}
-		bacon_video_widget_gst_codec_install_context_free (ctx);
-		return FALSE;
+	candidate = g_desktop_app_info_new ("eos-vlc.desktop");
+	if (candidate) {
+		message_text = _("This file type is currently unsupported in Videos. "
+		                 "However, you can use VLC media player "
+		                 "from the App Center to play it.");
+		return codec_missing_suggested_launch_action_new (message_text,
+								  _("Get VLC"),
+								  candidate,
+								  mrl);
 	}
 
-	return TRUE;
+	return NULL;
 }
 
 static void
 codec_confirmation_dialog_response_cb (GtkDialog       *dialog,
                                        GtkResponseType  response_type,
                                        gpointer         user_data)
 {
-	TotemCodecInstallContext *ctx = user_data;
+	GError *error = NULL;
+	CodecMissingSuggestedLaunchAction *action = user_data;
 
 	switch (response_type) {
 	case GTK_RESPONSE_ACCEPT:
-		bacon_video_widget_start_plugin_installation (ctx, FALSE);
+		if (action) {
```

Comment:
Technically you should only assign the action data in this block, since it's not passed through in the other code paths. At that point you could avoid the `NULL` check.